### PR TITLE
ARTEMIS-1326 testTimeOnTimedBuffer incorrect timeout count

### PR DESCRIPTION
**Pull request by franz1981**
The original test wasn't correctly counting the timeout expiration during the burst of writes.
Now the test is considering the timeout expirations between 2 consecutive bursts of writes in order to verify if the timeout is correctly handled.

### CODE REVIEW

**Comment by michaelandrepearce** [comment]
@franz1981 thanks for doing this

**Comment by clebertsuconic** [comment]
@franz1981 @michaelandrepearce the test was failing for a missing patch, right?

Was this test still failing sporadically?

**Comment by clebertsuconic** [comment]
The test was correct before.

This write here would flush the buffer right away and a new write would be issued:
https://github.com/apache/activemq-artemis/blob/f12116d5a54fbaa2f2507293bb6d524ec3012fa7/tests/unit-tests/src/test/java/org/apache/activemq/artemis/tests/unit/core/journal/impl/TimedBufferTest.java#L195-L196


The burst cannot complete in less than the timeout.. otherwise the timed buffer is not timing...

I would revert this change... or provide more info on where it failed.

**Comment by clebertsuconic** [comment]
The only reason it failed before was because of a non applied patch, and this test actually caught that.. you cannot change the test.

File: tests/unit-tests/src/test/java/org/apache/activemq/artemis/tests/unit/core/journal/impl/TimedBufferTest.java [hunk 1]
```diff
@@ -195,24 +194,22 @@ public void onError(int errorCode, String errorMessage) {
...
+         //Sending like crazy... still some wait (1 millisecond) between each send..
+         //The writes will be performed during a timeout interval (it can take < timeout time to do that!)
          long time = System.currentTimeMillis();
          for (int i = 0; i < 5; i++) {
             timedBuffer.addBytes(buff, true, callback);
             Thread.sleep(1);
          }
          Assert.assertTrue(latchFlushed.await(5, TimeUnit.SECONDS));
-
+         //The TimedBuffer will wait until timeout expiration to perform any additional write
```
Comment (clebertsuconic) [inline tests/unit-tests/src/test/java/org/apache/activemq/artemis/tests/unit/core/journal/impl/TimedBufferTest.java]: this was already done on this test.

**Comment by franz1981** [comment]
@clebertsuconic 

> The burst cannot complete in less than the timeout.. otherwise the timed buffer is not timing...

A burst can complete in less than a timeout (reading the callback calls in the test) but a new one can be accepted only when the timeout will be expired: that's what a sanity test should verify IMHO.
Logging the different states of the TimedBuffer in the old and the new version help to understand this little difference.
Reverting it would make it a step back from the correctness point of view IMHO and features equality between NIO/MAPPED/ASYNCIO, because now (after the last change you suggested me) it is really timing with the correct timeout...
@mtaylor  @michaelandrepearce have worked around these code paths too, wdyt?

**Comment by clebertsuconic** [comment]
@franz1981 no, a burst cannot complete in less than a timeout.. otherwise.. for the 10th, 11th time.. you would be breaking the TimedBuffer...


if Mapped can work without Timing, use a different buffer approach, don't break the TimedBuffer.

**Comment by clebertsuconic** [comment]
the point of the TimedBuffer is to limit the writes you can send to storage at a limit.... you write once.. flush... next time you write it must be in the configured timing...

the only reason why the flush could be flushed before time is if the buffer is full, and that is currently contemplated... anything else will be sending more writes (and possibly flushes) than what the device is capable of handling....


again.. if Mapped works differently, then don't use TimedBuffer.. but you can't change this semantic here.

**Comment by clebertsuconic** [comment]
>>but a new one can be accepted only when the timeout will be expired: <<

the test already had a previous write that could finish before the timeout... this one is already the new one.

I guess it wasn't clear on my test.. as there was no comment about that. so that was already part of the test.

**Comment by franz1981** [comment]
@clebertsuconic 
Probably we use the same term for different things: what do you mean by burst and complete?
For what I mean if you start calling addBytes several times until timeout is not finished yet and you have a flush method (months test observer) that do not sleep at least timeout time, it is normal that you will have your callbacks called in short time. 
What is important is that the buffer can't allow to flush new data until a new timeout will be expired: it is like the notion of credits for the flow control.
If you finish the credits immediately you can't have new ones until the timeout will finish and it is exactly what is the definition of IOPS for the disks.
It is not a matter of MAPPED or not...but a matter of a test that is not emulations any correct disk behaviour: asynchronously o not.

**Comment by michaelandrepearce** [comment]
Hi @franz1981,  @clebertsuconic has a good point, the issue i had was because i hadn't rebased actually, and since rebasing i hadn't seen the issue again on PR builds.

**Comment by clebertsuconic** [comment]
@franz1981 the test was correct before, there's no reason to make this PR

**Comment by clebertsuconic** [comment]
@franz1981 can you please close this PR, and leave this test as is?

**Comment by franz1981** [comment]
I'll close this because seems fixed by: https://github.com/apache/activemq-artemis/commit/27d37b735d2248a6ce81f0a6796e26c1a986cb93

**Pull request closed without merging**